benches/impl_path_string_for_evaluation_context.rs

```diff
@@ -32,47 +33,45 @@ const HEADERS_VALUE: &[&[&str]] = &[&["headers", "existing"], &["headers", "miss
 
 const VARS_VALUE: &[&[&str]] = &[&["vars", "existing"], &["vars", "missing"]];
 
-lazy_static::lazy_static! {
-  static ref TEST_VALUES: Value = {
-    let mut root = IndexMap::new();
-    let mut nested = IndexMap::new();
+static TEST_VALUES: Lazy<Value> = Lazy::new(|| {
```

Comment:
Can you run and showcase the difference in performance that's produced via these benchmarks.

Reply:
Sure, I will run the benchmarks.

Reply:
Previous benchmark results for the commit b4196a654c3bb235d265bdb18ee2d6ad19e88cce
```bash
test result: ok. 0 passed; 0 failed; 125 ignored; 0 measured; 0 filtered out; finished in 0.02s

     Running unittests src/main.rs (target/release/deps/tailcall-291086a3e683fcad)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/data_loader_bench.rs (target/release/deps/data_loader_bench-d0d578e69e12ee73)
Gnuplot not found, using plotters backend
test_data_loader        time:   [773.16 µs 775.00 µs 776.93 µs]
                        change: [-3.8791% -2.9808% -2.1910%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  2 (2.00%) high severe

     Running benches/impl_path_string_for_evaluation_context.rs (target/release/deps/impl_path_string_for_evaluation_context-d4184ab936dc0754)
Gnuplot not found, using plotters backend
input/value.root        time:   [50.446 ns 50.520 ns 50.593 ns]

input/value.nested.existing
                        time:   [97.545 ns 97.642 ns 97.737 ns]
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low severe
  3 (3.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe

input/value.missing     time:   [44.254 ns 44.283 ns 44.314 ns]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild

input/value.nested.missing
                        time:   [86.531 ns 86.807 ns 87.109 ns]
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

input/args.root         time:   [23.301 ns 23.329 ns 23.360 ns]

input/args.nested.existing
                        time:   [73.268 ns 73.335 ns 73.401 ns]
Found 12 outliers among 100 measurements (12.00%)
  12 (12.00%) low mild

input/args.missing      time:   [17.363 ns 17.370 ns 17.378 ns]
Found 11 outliers among 100 measurements (11.00%)
  9 (9.00%) high mild
  2 (2.00%) high severe

input/args.nested.missing
                        time:   [60.230 ns 60.349 ns 60.516 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

input/headers.existing  time:   [24.279 ns 24.331 ns 24.379 ns]

input/headers.missing   time:   [24.091 ns 24.119 ns 24.147 ns]
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) low severe
  2 (2.00%) low mild
  2 (2.00%) high mild

input/vars.existing     time:   [7.1270 ns 7.1484 ns 7.1867 ns]
Found 13 outliers among 100 measurements (13.00%)
  7 (7.00%) high mild
  6 (6.00%) high severe

input/vars.missing      time:   [7.1071 ns 7.1250 ns 7.1413 ns]

     Running benches/json_like_bench.rs (target/release/deps/json_like_bench-2e7385627799dfec)
Gnuplot not found, using plotters backend
test_batched_body       time:   [1.4497 µs 1.4521 µs 1.4545 µs]
Found 7 outliers among 100 measurements (7.00%)
  7 (7.00%) high mild

     Running benches/request_template_bench.rs (target/release/deps/request_template_bench-8c32648d9161deb6)
Gnuplot not found, using plotters backend
with_mustache_literal   time:   [676.79 ns 677.97 ns 679.30 ns]
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) high mild

with_mustache_expressions
                        time:   [1.1461 µs 1.1481 µs 1.1500 µs]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
  ```
  
  Here are the result for the commit addressing the issue. ac8a2eb53777ee90d414ad5d6cbbe0b40c870962
  
  ```bash
  Benchmarking test_data_loader: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 5.4s, enable flat sampling, or reduce sample count to 60.
test_data_loader        time:   [990.91 µs 995.89 µs 1.0015 ms]
                        change: [-3.0511% -2.1860% -1.1858%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

     Running benches\impl_path_string_for_evaluation_context.rs (target\release\deps\impl_path_string_for_evaluation_context-a25576a90835e87f.exe)
Gnuplot not found, using plotters backend
input/value.root        time:   [69.128 ns 69.406 ns 69.707 ns]
                        change: [+0.1397% +0.6154% +1.1939%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

input/value.nested.existing
                        time:   [133.69 ns 134.04 ns 134.40 ns]
                        change: [+1.3500% +1.6728% +2.0232%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild
  2 (2.00%) high severe

input/value.missing     time:   [64.139 ns 64.507 ns 64.952 ns]
                        change: [-2.8853% -2.1377% -1.2819%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  10 (10.00%) high mild

input/value.nested.missing
                        time:   [127.64 ns 128.10 ns 128.56 ns]
                        change: [-1.0285% -0.6286% -0.2338%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) high mild

input/args.root         time:   [27.592 ns 27.620 ns 27.651 ns]
                        change: [+0.5180% +0.7338% +0.9559%] (p = 0.00 < 0.05)
                        Change within noise threshold.

input/args.nested.existing
                        time:   [89.646 ns 90.051 ns 90.499 ns]
                        change: [-2.2413% -1.8755% -1.5169%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

input/args.missing      time:   [17.418 ns 17.428 ns 17.441 ns]
                        change: [+0.7144% +0.8338% +0.9531%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 17 outliers among 100 measurements (17.00%)
  5 (5.00%) low mild
  10 (10.00%) high mild
  2 (2.00%) high severe

input/args.nested.missing
                        time:   [88.976 ns 89.306 ns 89.617 ns]
                        change: [-0.3172% +0.1472% +0.6007%] (p = 0.54 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  6 (6.00%) low mild
  1 (1.00%) high mild

input/headers.existing  time:   [26.306 ns 26.326 ns 26.351 ns]
                        change: [+1.1853% +1.3084% +1.4196%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  7 (7.00%) high mild
  4 (4.00%) high severe

input/headers.missing   time:   [26.667 ns 26.718 ns 26.774 ns]
                        change: [-0.0796% +0.2386% +0.5853%] (p = 0.18 > 0.05)
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild

input/vars.existing     time:   [8.0280 ns 8.0403 ns 8.0540 ns]
                        change: [+0.8080% +1.0365% +1.2461%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

input/vars.missing      time:   [6.4088 ns 6.4202 ns 6.4335 ns]
                        change: [+1.4931% +1.7688% +2.0405%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  9 (9.00%) high mild

     Running benches\json_like_bench.rs (target\release\deps\json_like_bench-bf62a49c361bbe2f.exe)
Gnuplot not found, using plotters backend
test_batched_body       time:   [4.2292 µs 4.2452 µs 4.2654 µs]
                        change: [-1.1492% -0.4701% +0.1447%] (p = 0.17 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  9 (9.00%) high severe

     Running benches\request_template_bench.rs (target\release\deps\request_template_bench-58945be0dfeac3c9.exe)
Gnuplot not found, using plotters backend
with_mustache_literal   time:   [960.94 ns 962.67 ns 964.74 ns]
                        change: [+1.7459% +2.0161% +2.3019%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

with_mustache_expressions
                        time:   [1.7028 µs 1.7054 µs 1.7080 µs]
                        change: [+1.4397% +1.6382% +1.8477%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
```

Reply:
This is some significant degradation in performance. Can you explain that what's happening?

Reply:
> Can you run and showcase the difference in performance that's produced via these benchmarks.

There is one issue I found with benchmarking procedure, so previously I ran cargo bench for commit b4196a6 on wsl2 ubuntu 20LTS whereas the benchmarks for commit 8c8a2ab was done on windows 10 x86.

and I did the same test for commit b4196a6 on both the platforms and results were remarkably different ( perhaps due to better translation layer of linux) , I didn't knew the exact cause for this but perhaps this can be useful reference linked [here](https://users.rust-lang.org/t/effect-on-benchmarking-due-to-platform/102088)  

After knowing the issue I ran fresh test for commit b4196a6 on WSL2 ubuntu 20LTS

```rust
     Running benches/data_loader_bench.rs (target/release/deps/data_loader_bench-d0d578e69e12ee73)
Gnuplot not found, using plotters backend
test_data_loader        time:   [781.11 µs 782.38 µs 783.77 µs]
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe

     Running benches/impl_path_string_for_evaluation_context.rs (target/release/deps/impl_path_string_for_evaluation_context-d4184ab936dc0754)
Gnuplot not found, using plotters backend
input/value.root        time:   [42.461 ns 42.543 ns 42.626 ns]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

input/value.nested.existing
                        time:   [80.378 ns 80.524 ns 80.656 ns]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild

input/value.missing     time:   [44.066 ns 44.094 ns 44.127 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

input/value.nested.missing
                        time:   [75.609 ns 75.688 ns 75.773 ns]
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) low mild
  4 (4.00%) high mild

input/args.root         time:   [23.318 ns 23.327 ns 23.336 ns]
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe

input/args.nested.existing
                        time:   [66.205 ns 66.298 ns 66.432 ns]
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe

input/args.missing      time:   [20.143 ns 20.172 ns 20.202 ns]
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) low severe
  2 (2.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe

input/args.nested.missing
                        time:   [52.397 ns 52.630 ns 52.953 ns]
Found 16 outliers among 100 measurements (16.00%)
  2 (2.00%) low mild
  14 (14.00%) high severe

input/headers.existing  time:   [24.323 ns 24.353 ns 24.380 ns]

input/headers.missing   time:   [23.877 ns 23.903 ns 23.932 ns]
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe

input/vars.existing     time:   [7.1168 ns 7.1282 ns 7.1402 ns]
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

input/vars.missing      time:   [7.0681 ns 7.0850 ns 7.1005 ns]

     Running benches/json_like_bench.rs (target/release/deps/json_like_bench-2e7385627799dfec)
Gnuplot not found, using plotters backend
test_batched_body       time:   [1.4325 µs 1.4340 µs 1.4357 µs]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild

     Running benches/request_template_bench.rs (target/release/deps/request_template_bench-8c32648d9161deb6)
Gnuplot not found, using plotters backend
with_mustache_literal   time:   [671.78 ns 673.02 ns 674.16 ns]
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) low mild
  3 (3.00%) high mild

with_mustache_expressions
                        time:   [1.1100 µs 1.1121 µs 1.1143 µs]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  ```
  and did the same for commit ac8a2eb
  
  ```rust
       Running benches/data_loader_bench.rs (target/release/deps/data_loader_bench-b425ffdf80c246df)
Gnuplot not found, using plotters backend
test_data_loader        time:   [762.41 µs 765.16 µs 767.98 µs]
                        change: [-1.4903% -0.9992% -0.4808%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

     Running benches/impl_path_string_for_evaluation_context.rs (target/release/deps/impl_path_string_for_evaluation_context-1d1fdc65bcdb2c96)
Gnuplot not found, using plotters backend
input/value.root        time:   [41.656 ns 41.739 ns 41.826 ns]
                        change: [-2.0459% -1.7402% -1.4198%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

input/value.nested.existing
                        time:   [81.166 ns 81.373 ns 81.561 ns]
                        change: [-1.0599% -0.5053% +0.0694%] (p = 0.08 > 0.05)
                        No change in performance detected.

input/value.missing     time:   [35.664 ns 35.712 ns 35.757 ns]
                        change: [-19.360% -19.146% -18.852%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high severe

input/value.nested.missing
                        time:   [71.264 ns 71.403 ns 71.537 ns]
                        change: [-6.4788% -6.2335% -6.0036%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) low severe
  1 (1.00%) low mild
  1 (1.00%) high mild

input/args.root         time:   [22.666 ns 22.685 ns 22.707 ns]
                        change: [-3.2770% -2.9219% -2.6965%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  5 (5.00%) low severe
  4 (4.00%) high mild
  6 (6.00%) high severe

input/args.nested.existing
                        time:   [61.721 ns 61.797 ns 61.872 ns]
                        change: [-7.4425% -7.0492% -6.7046%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe

input/args.missing      time:   [17.024 ns 17.034 ns 17.045 ns]
                        change: [-15.745% -15.589% -15.305%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low severe
  5 (5.00%) high mild
  1 (1.00%) high severe

input/args.nested.missing
                        time:   [52.098 ns 52.231 ns 52.381 ns]
                        change: [-1.2146% -0.6730% -0.1964%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

input/headers.existing  time:   [24.335 ns 24.356 ns 24.379 ns]
                        change: [+0.1977% +0.3967% +0.5879%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

input/headers.missing   time:   [23.114 ns 23.160 ns 23.212 ns]
                        change: [-3.1893% -2.9245% -2.6571%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe

input/vars.existing     time:   [6.9740 ns 6.9785 ns 6.9839 ns]
                        change: [-2.4191% -2.0222% -1.7488%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

input/vars.missing      time:   [6.7551 ns 6.7619 ns 6.7695 ns]
                        change: [-4.6534% -4.4583% -4.2622%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) high mild

     Running benches/json_like_bench.rs (target/release/deps/json_like_bench-b3c9edd104666420)
Gnuplot not found, using plotters backend
test_batched_body       time:   [1.4555 µs 1.4565 µs 1.4574 µs]
                        change: [+1.2264% +1.3522% +1.4780%] (p = 0.00 < 0.05)
                        Performance has regressed.

     Running benches/request_template_bench.rs (target/release/deps/request_template_bench-b095d634fd6436e8)
Gnuplot not found, using plotters backend
with_mustache_literal   time:   [675.02 ns 676.43 ns 677.85 ns]
                        change: [+0.3564% +0.6454% +0.9374%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild

with_mustache_expressions
                        time:   [1.1385 µs 1.1408 µs 1.1430 µs]
                        change: [+2.1099% +2.3344% +2.5647%] (p = 0.00 < 0.05)
                        Performance has regressed.
```
So as per the results shown here I feel we are in good spot, also I would urge @tusharmath to please confirm the results on your machine too 😊.